### ISSUE TITLE
Use factories and not fixtures

### ISSUE THREAD

**Issue by andreareginato** [andreareginato/betterspecs#11]
Write your thoughts about the "[use factories and not fixtures](http://betterspecs.org/#factories)" best practice.

**Comment by nruth** [comment]
> The kinds of functions I am talking about are things like functions called create_user with crazy levels of arguments to define how the user factory instance is made 

Yes this does happen, e.g. wrapping the must-be-in-one-transaction-so-cant-be-belongs-to shortcoming of old tools. Anyway, it seems the document should be updated to say "Don't use <s>Rails helpers</s> spec/support helpers", instead of don't use fixtures. 

In fact, for complex setup cases [fixtures sound pretty good](https://whatdoitest.com/getting-friendly-with-fixtures) and I'm going to give them a spin for climbing out of a 4-7 year old hole of overlapping helper mess & slow factories.

Beyond that it'd mean showing sustainable ways to use both approaches, which is probably no good for a pithy guide like this.

**Comment by marnen** [comment]
@nruth That article you linked to drives me crazy, because it seems to go out of its way to show some of the ways that fixtures suck, but then unaccountably says "oh, they're not so bad"!

They are so bad; they encourage inflexible test data that's overly separated from your test functions. What you probably want instead is more factory templates to easily cover different cases.

**Comment by ttilberg** [comment]
It's been five years of discussion, and the rule in the guide still makes a pretty dramatic claim about fixtures without stating why. The example given has nothing to do with fixtures. I've recently been trying to research the reasons people flatly say "don't use fixtures" but in every case I've come across, they are used in an arduous way that isn't necessary. If the example were actually using fixtures, it would look like this:

> Use factories and not fixtures
> 
> This is an old topic, but it's still good to remember it. Do not use fixtures because they are difficult to control, use factories instead. Use them to reduce the verbosity on creating new data.
> 
> BAD
> user = users(:guy)
> 
> GOOD
> user = FactoryGirl.create :user

Well, that's not really compelling at all. In the published example, you aren't even using fixtures, so of course it's more verbose. A common argument is often that you have to maintain fixtures and give them default values and whatnot. But... don't you also do that with FactoryGirl?

Being that the fixtures can also include generated data if you really need it to (such as `name: <%= Faker::Name.first_name %>`, or create 1000 random objects in quick loop), I struggle to understand the argument here. What's difficult to control about sourcing an object with default values, and when necessary, changing them to fit your test using things like `user = users(:tim); user.name = 'special name for this case'` Even complex associations can either be saved in the Fixture, or composed on the fly in your test, just like in FactoryGirl.

I fully admit that I don't have a lot of experience with rspec and FactoryGirl, as I've found MiniTest and fixtures very sufficient in my few years of study. It's very possible that my lack of experience in huge apps has me blind to something. That said, the motivation of my comment here is the search for that experience -- but everything I've found so far has been extremely unconvincing.

**Comment by marnen** [comment]
@ttilberg:

>  A common argument is often that you have to maintain fixtures and give them default values and whatnot. But... don't you also do that with FactoryGirl?

Not in the same way. A factory is a *template*, with default values that you're expected to override as necessary when you call it. A fixture, OTOH, is a fixed record that you're expected to fetch as is. (At least, this was the case when I last used fixtures, admittedly a long time ago.)

So with fixtures, you define your test data far away from your test functions, and you need to predeclare every test record. With factories, you define your test data *right in the test*, and you do so on the fly without much predeclaration. The difference may look small at the beginning, but gets more apparent as the tests get more complex. 

Predeclaring is time-consuming and takes you out of the creative flow when you're writing your test. If you need a different case of test record with fixtures, you'll be tempted to reuse an existing one because of the overhead of predeclaring, so your test data won't be sufficiently specific to actually test what you think you're testing. If you need a different case with factories, there is no overhead; you just say exactly what you want, so you can test exactly what you think you're testing. 


> What's difficult to control about sourcing an object with default values, and when necessary, changing them to fit your test

Nothing as stated. But if you really wanted a user named John, why did you have to first create Tim and then change his name? Something like Factory Girl would let you just create John right at the beginning without declaring a new template or fixture. In your example, there's not a huge difference, but in a system that had an audit log for each change to a record, there would be a big difference, and it could mess up your tests. 

Moreover, if you only have a few fixtures and you're changing their data for every test, then you have no reason *not* to use factories: that is more or less the usage pattern that Factory Girl was designed for, and it provides a nicer interface for it. 

> Even complex associations can either be saved in the Fixture, or composed on the fly in your test, just like in FactoryGirl.

Saving a complex association in the fixture is usually bad, because it makes your test data too rigid. Composing on the fly without syntactic help is awkward as soon as there's more than one level of association. Factory Girl strikes a happy medium where the association *pattern* is assisted, but the association *data* is flexible on the fly. 

> I fully admit that I don't have a lot of experience with rspec and FactoryGirl, as I've found MiniTest and fixtures very sufficient in my few years of study.

If you don't have a lot of experience with factories, then why are you making sweeping statements about them? (Also, check out RSpec and/or MiniTest::Spec: the syntax is more readable and helps you focus your tests on interface, not implementation.)

Anyway, this sounds like classic Blub Paradox, but I know that's not an especially satisfactory answer. So: for me, a big reason that I like factories is that I can create exactly the test data I need, right where I need it, without predeclaring or confusing the test data. 

I'll grant you that the differences may initially seem subtle and may not be immediately apparent to a novice. But they grow bigger as projects get more complex and as tests get more complete. 

I'd go so far as to say that if you find fixtures sufficient, you probably fall into one of three cases:
* Your project is trivial (so it doesn't matter which method you use right now, but your project *will* grow)
* You're not testing well enough (because fixtures make it *very* hard to do so)
* You're abusing fixtures by treating them like factories as suggested above (in which case, why not use factories?)

In closing, a thought: some people who switch from fixtures to factories try to use factories like fixtures by predeclaring every fixture as a factory template. If you're doing that, stop now: that's not how factories are meant to be used. 

***

If you have sample test code available, I would be happy to have a look at how you use fixtures and see what I can tell you about the differences *for your use case*.

**Comment by ttilberg** [comment]
I appreciate the reply, and all points are taken with consideration. You're right that my apps are mostly small data management applications for ETL and client management -- fixtures work great. But I want to correct this part as it's directly related to this issue, and my original comment:

> If you don't have a lot of experience with factories, then why are you making sweeping statements about them? 

That's actually the exact opposite of what's going on here. I'm not making any assumptions at all about rspec/minitest, or criticizing them. In fact, I'm fully admitting I don't know them well. This is how I ended up on this project's page in the first place, and found this exact issue to discuss this point in the project. 

When you are learning rails, minitest and fixtures are what's included -- so that's what you learn.

The project is making the sweeping statement, asserting that fixtures are bad and you should opt for factories without helping a user understand why. The given example has nothing to do with fixtures.

My goal with these comments isn't to criticize or argue about the merits of Fixtures v FactoryGirl, but to bring to light that for someone who uses fixtures regularly, [the content of the subject at hand](http://www.betterspecs.org/#factories) simply doesn't make sense. It makes a dramatic statement ("don't use the thing that comes bundled with Rails and is well documented in the official documentation") while intentionally misrepresenting the example.

So, like I mentioned -- I don't know enough about this subject to be able to write a PR for this to improve it. I just know that as it stands, this doesn't make much sense, and this is the forum for discussing that.

**Comment by marnen** [comment]
Ah, good point; the example on the betterspecs website is terrible. I was responding more to the general principle than to the website content. 

> When you are learning rails, minitest and fixtures are what's included -- so that's what you learn.

Yes, and then you quickly discard them when you realize how much they suck. :) FWIW, I've worked as a Rails developer since about 2008, at a variety of companies. I think every one of them has used RSpec and factories. They are kind of community defaults at this point, even if the Rails core team doesn't want to distribute them with the Rails gem. :)

Now, I'm not recommending them *because* they're widely used, but rather because I think they're better than what comes standard with Rails. But I do think their near-universal acceptance is an interesting point.

**Comment by redbar0n** [comment]
I think this illustrates the pain point with fixtures pretty well.

TL;DR: Changing a fixture, or even adding a new fixture will often break a vast number of previous tests*, and fixtures introduces mystery guests.

\* \- Sometimes you need to test methods that f.ex. outputs statistic based on the DB state. Even adding 1 new fixture later on, would break all those tests of methods relying on counting records in the DB. Maybe there's a workaround (?), but that is what I am experiencing.

> For data sets that are a bit more involved, YAML fixtures are difficult to maintain and hard to change without affecting other tests. I mean, you can make them work, of course—after all, developers used them plenty in the past—but tons of developers will agree that the price to pay for managing fixtures is just a bit stingy.
> 
> One scenario we definitely want to avoid is changing little details on an existing fixture and causing tons of tests to fail. If these failing tests are unrelated, the situation is even worse—a good example of tests being too brittle. In order to “protect” existing tests from this scenario, this can also lead to growing your fixture set beyond any reasonable size—being DRY with fixtures is most likely not on the table anymore at that point. 
> 
> To avoid breaking your test data when the inevitable changes occur, developers were happy to adopt newer strategies that offered more flexibility and dynamic behaviour. That’s where Factory Girl came in and kissed the YAML days goodbye. 
> 
> Another issue is the heavy dependency between the test and the .yml fixture file. Since the fixtures are defined in a separate .yml file, mystery guests are also a major pain waiting to bite you due to being obscure.

Source: [https://code.tutsplus.com/articles/antipatterns-basics-rails-tests--cms-26011](https://code.tutsplus.com/articles/antipatterns-basics-rails-tests--cms-26011)

It also describes what Mystery Guests are, and the problem with them (they are akin to a [mode, and it's the problems they entail](https://itsthedatastupid.wordpress.com/2010/05/29/no-modes/)). One counter-argument is that mystery guests is a result of stringently applying Convention over Configuration.

That said, I see the benefits of fixtures as well, and I'm torn between that and the admittedly more sexy FactoryGirl.

I tried finding an elaboration on DHH's stance on it, but the most detailed I could find was this:

> Use the power of testing fixtures and the case-transaction loop to avoid having to do expensive and excessive setup work for every case, and merely tailor your prototypical fixtures to the case at hand by changing the single state needed. (More on this technique in that promised future post).

Source: [http://david.heinemeierhansson.com/2014/test-induced-design-damage.html](http://david.heinemeierhansson.com/2014/test-induced-design-damage.html)

I couldn't find that promised future post. Let me know if any of you find it.

The case-transaction loop he is referring to (a novel expression, I believe), is probably [ActionModel::Base.transaction](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html).
By _"tailor your prototypical fixtures to the case at hand by changing the single state needed"_, I presume he means using `Model.update_attributes` (which will update the record in the DB which the YAML fixtures had ensured was already there.) in each test (or in the setup method).

I think this is doing exactly what @marnen advised against above:
_"You're abusing fixtures by treating them like factories as suggested above (in which case, why not use factories?)"_ But to DHH that seems to be the golden standard, perhaps oddly enough. Maybe you can actually get the best of both worlds with this approach? Using a complete set of fully fleshed out fixtures (using real world data), in all their complexity and interconnectedness, while overriding in the few cases where it is necessary..

I think the whole issue with the betterspec page though, is that it uses "fixtures" in an ambiguous way: "fixtures" in the broad sense which means pre-created fixed data (like `Model.create`), and not "fixtures" in the narrow sense that is common in Rails, which refers to the YAML files approach. The page should probably be updated to remove this ambiguity.

I second the suggestion by @dgm _"What I want is a VCR like gem for databases."_ to try to get the best from both factories and fixtures and the worst of neither.

(If I'm allowed to dream, outside of this discussion, I would also want at VCR-like gem that would allow a non-developer to click through a use case in the app web interface, and store that as a test case which then capybara or similar could run. Automatically generated/macroed integration tests. Sorry for digressing, I just wanted that idea to be out there.)

**Comment by marnen** [comment]
@redbar0n FWIW, I no longer pay much attention to how DHH wants me to use Rails. Of course he did fantastic work in creating (the early versions of) the framework, but perhaps because of that, he's become so attached to his earlier vision that he can't seem to understand that there are other (IMHO better) ways of doing things than he envisioned. It's got to the point where by default, I take his recommendations as a slight _negative_ these days.

**Comment by redbar0n** [comment]
@marnen What benefits does fixtures have?

"We avoid file system, database, simply because these 'shared fixture' elements prevent us running in isolation from other tests, or cause our tests to be slow." - Ian Cooper (36:42) https://youtu.be/EZ05e7EMOLM?t=2202

This was rebutted by DHH in https://dhh.dk/2014/slow-database-test-fallacy.html since DB runs can be fast, and fixtures can be run in isolation. 

But isolation needs not only to happen at runtime, but the data itself need to be isolated, so that changing the fixture for one test doesn't inadvertently break another test. Shared fixtures introduce coupling, as mentioned in the previous post.

I think this tweet by dhh illustrates how he circumvents that problem:

"Only write a handful of fixtures per class to spin up the world. Then mutate base fixtures as required per test case." - dhh on twitter: https://twitter.com/dhh/status/449673835353477120

Then there are some benefits to fixtures, since you can describe a general "world" once, which can be shared between test cases without actually introducing any meaningful coupling. Because whenever you change anything in that world which is not shared amongst all test cases, then you use a small mutation in the diverging test case in question. Then you achieve Kent Beck's principle that all tests are isolated from other tests.

("Kent uses 'unit test' in a very specific way. When he uses the phrase, all he means is that: You should be able to run all tests together, in one suite, without one test being run impacting the other test. The unit of isolation is the test. A lot of mistakes has been made, because people think that the unit of isolation is the class under test. It is not." - Ian Cooper, 35:00, https://www.youtube.com/watch?v=EZ05e7EMOLM )

The problem with this approach is, of course, that if not all developers are on the same page, someone will change the base fixtures themselves, instead of mutating them in their test, and then crash a bunch of other tests. Hopefully this should be caught at inception when tests are immediately run afterwards, and by inferring the principle dhh tweeted from how other developers wrote their tests.

**Comment by marnen** [comment]
@redbar0n I think you might possibly have missed my point. It seems like you’ve mostly listed advantages of testing with fixtures as opposed to no DB management. My claim, though, was about fixtures *as opposed to factories*—and there I claim that anything that fixtures do, factories do better. Your comment mostly doesn’t seem to address this, unless I’m misreading. 

***

> Shared fixtures introduce coupling, as mentioned in the previous post.

> Then there are some benefits to fixtures, since you can describe a general "world" once, which can be shared between test cases without actually introducing any meaningful coupling. 

These two points are mutually contradictory. You can’t really do this with shared fixtures without introducing coupling between tests. The right way to do this—if you really must—is to *describe* the world once, but to *create* it for each test. This is what factories in an RSpec `before` block would do, for example. 

But I’m not sure it’s a good idea anyway. I find that a general “world” is far too much setup and far too many assumptions for unit testing. I prefer to customize the world for each test case. Sure, there will be some
common aspects between cases, but I generally want as little of that as I can get away with. That way I’m making as few assumptions as possible for each test. And that’s much easier to do with factories than with fixtures, because it’s easier to use factories dynamically. 

> The problem with this approach is, of course, that if not all developers are on the same page, someone will change the base fixtures themselves, instead of mutating them in their test, and then crash a bunch of other tests. 

Right. Why do that to yourself in the first place? It gives no compensating advantage and just introduces the temptation to bring a lot of unnecessary garbage into your test setup.

**Comment by marnen** [comment]
@redbar0n:

> I would also want at VCR-like gem that would allow a non-developer to click through a use case in the app web interface, and store that as a test case which then capybara or similar could run.

This is not the topic for that, but: maybe https://www.selenium.dev/selenium-ide/ ? Or Cucumber with a rich set of step definitions?

**Comment by redbar0n** [comment]
@marnen Actually, I was thinking fixtures vs. factories myself when I wrote it. It is my prime motivation here to find out the important differences, and try to deduce which is better when. DHH's tweet was particularly elucidating on how he uses fixtures successfully. In that way you can actually use fixtures without meaningful coupling, because the fixtures are actually "fixed", and only local copies of them in each tests are mutated. The "world" you describe with fixtures would be equivalent to the factories way of using shared instantiations in the "before test hooks" (particularly [`before(:suite)` in rspec](https://relishapp.com/rspec/rspec-core/v/2-0/docs/hooks/before-and-after-hooks#define-before(:all)-block-in-example-group)) at the top of the test file. In both cases these are then mutated only in each test (and [in case of fixtures then mutations are automatically rolled back after each test](https://dhh.dk/2014/slow-database-test-fallacy.html)). And in both cases (fixtures, or with factories using the `before` test hooks) there is coupling, but not 'meaningful coupling' in the sense that changing something inside one test will break another test (the unit of isolation is the test). I think we are very much on the same page in preferring to customise the world for each test case. My point was that this is even what DHH is doing with fixtures. Which avoids how people typically misuse fixtures when coming into them: making single-test-modifications inside the global fixtures. Instead of just making changes to the preloaded models inside each test like DHH seems to do.

I have found that creating a general "world" can be easier with fixtures than with factories. Because with fixtures that world could be a simple copy of your production DB, if you'd like. Or you could create it manually, but with seemingly real-world data. The benefit to the programmer testing is that he/she would only have to know "one world" really well, and be able to reason about it quite easily, to be able to write tests. Instead of when using factories typically creating "many small worlds", with typically bogus data, "just to get a that model up and running". Of course, you could create "one world" with factories as well, but then it would typically be spread out over several factory instantiation files.

>Why do that to yourself in the first place? It gives no compensating advantage and just introduces the temptation to bring a lot of unnecessary garbage into your test setup.

All the above said, I agree with you that it's easier to use factories dynamically, and to keep your test surface lean and small for each individual test, so I see a lot of good reasons to use them. My only point was that the issue was not as clear-cut as I initially thought it to be. DHH is heavily on the side of pragmatism and for least-effort-approaches (the ethos of RoR), so it carried some weight with me that he was such a stark proponent for sticking with fixtures (and using them in the right way, which the rest of us might have simply missed from the get-go).

More on DHH's bottom line suggestions for testing Rails/MVC apps: https://dhh.dk/2014/test-induced-design-damage.html

**Comment by redbar0n** [comment]
> @redbar0n:
> 
> > I would also want at VCR-like gem that would allow a non-developer to click through a use case in the app web interface, and store that as a test case which then capybara or similar could run.
> 
> This is not the topic for that, but: maybe https://www.selenium.dev/selenium-ide/ ? Or Cucumber with a rich set of step definitions?

Thanks for the tip. :) I knew about Selenium and Cucumber, but Selenium IDE seems to be close to my idea! Sadly, [Selenium IDE is deprecated](https://seleniumhq.wordpress.com/2017/08/09/firefox-55-and-selenium-ide/), but other tools such as [applitools.com](https://applitools.com) and [katalon.com](https://www.katalon.com/) has taken its place, apparently. Actually [Katalon Automation Recorder](https://www.katalon.com/resources-center/blog/selenium-ide-alternative-firefox-chrome/) seems to be exactly what I was thinking about! <3
